contracts/Vote.sol

```diff
@@ -0,0 +1,427 @@
+        proposalVoteRecord.abstainVotes -= previousVoterRecord.abstainVotes;
+
+        // Add new vote.
+        proposalVoteRecord.yesVotes += yesVotes;
+        proposalVoteRecord.noVotes += noVotes;
+        proposalVoteRecord.abstainVotes += abstainVotes;
+
+        voteRecords[proposalId] = ProposalVoteRecord(
+            proposalId,
+            proposalVoteRecord.yesVotes,
+            proposalVoteRecord.noVotes,
+            proposalVoteRecord.abstainVotes
+        );
+
+        if (previousVoterRecord.proposalId == 0) {
+            voter.votedProposalIds.push(proposalId);
+        }
+
+        voter.proposalVotes[proposalId] = VoterRecord(proposalId, yesVotes, noVotes, abstainVotes);
+
+        emit ProposalVoted(accountVoter, proposalId, yesVotes, noVotes, abstainVotes);
+
+        if (proposalTimestamps[proposalId] == 0) {
+            proposalTimestamps[proposalId] = getProposalTimestamp(proposalId);
+        }
+
+        return (
+            toStakedCelo(totalWeights),
+            proposalVoteRecord.yesVotes,
+            proposalVoteRecord.noVotes,
+            proposalVoteRecord.abstainVotes
+        );
+    }
+
+    /**
+     * @notice Revokes votes on already voted proposal.
+     * @param accountVoter The account that is voting.
+     * @param proposalId The ID of the proposal to vote on.
+     * @return totalYesVotes SUM of all AccountContract yes votes for proposal.
+     * @return totalNoVotes SUM of all AccountContract no votes for proposal.
+     * @return totalAbstainVotes SUM of all AccountContract abstain votes for proposal.
+     */
+    function revokeVotes(address accountVoter, uint256 proposalId)
+        public
+        onlyManager
+        returns (
+            uint256 totalYesVotes,
+            uint256 totalNoVotes,
+            uint256 totalAbstainVotes
+        )
+    {
+        (, totalYesVotes, totalNoVotes, totalAbstainVotes) = voteProposal(
+            accountVoter,
+            proposalId,
+            0,
+            0,
+            0
+        );
+        return (totalYesVotes, totalNoVotes, totalAbstainVotes);
+    }
+
+    /**
+     * @notice Returns save timestamp of proposal.
+     * @param proposalId The proposal UUID.
+     * @return The timestamp of proposal.
+     */
+    function getProposalTimestamp(uint256 proposalId) public view returns (uint256) {
+        (, , uint256 timestamp, , ) = getGovernance().getProposal(proposalId);
+        return timestamp;
+    }
+
+    /**
+     * @notice Retuns currently locked celo in voting. (This celo cannot be unlocked.)
+     * And it will remove voted proposals from account history if appropriate.
+     * @param beneficiary The beneficiary.
+     */
+    function updateHistoryAndReturnLockedStCeloInVoting(address beneficiary)
+        public
+        onlyManager
+        returns (uint256 lockedAmount)
+    {
+        Voter storage voter = voters[beneficiary];
+
+        uint256 i = voter.votedProposalIds.length;
+        while (i > 0) {
```

Comment:
If a user had been voting for many proposal and forgot to call this function for a long time, this function could run out gas, no? Potential fix: have a version of this function that takes a parameter to limit the number of iterations (and we'd probably have to store the current max vote weight per user for that to work?).

Reply:
If this function will run out of gas there is still deleteExpiredVoterProposalId function that allows to delete user voted proposals one by one